File: sdk/python/src/dovewai_protocol/v02.py
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any
from uuid import uuid4

PROTOCOL_VERSION = "0.2"


def _now() -> str:
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")


def _base(object_type: str, object_id: str | None = None, created_at: str | None = None) -> dict[str, Any]:
    return {
        "protocol_version": PROTOCOL_VERSION,
        "id": object_id or str(uuid4()),
        "type": object_type,
        "created_at": created_at or _now(),
    }


def task(intent: str, **fields: Any) -> dict[str, Any]:
    return {**_base("Task", fields.pop("id", None), fields.pop("created_at", None)), "intent": intent, **fields}


def capability(name: str, version: str, **fields: Any) -> dict[str, Any]:
    return {**_base("Capability", fields.pop("id", None), fields.pop("created_at", None)), "name": name, "version": version, **fields}


def claim(task_id: str, holder_id: str, mode: str, lease_expires_at: str, **fields: Any) -> dict[str, Any]:
    return {
        **_base("Claim", fields.pop("id", None), fields.pop("created_at", None)),
        "task_id": task_id,
        "holder_id": holder_id,
        "mode": mode,
        "lease_expires_at": lease_expires_at,
        **fields,
    }


def attempt(task_id: str, attempt_number: int, status: str, **fields: Any) -> dict[str, Any]:
    return {
        **_base("Attempt", fields.pop("id", None), fields.pop("created_at", None)),
        "task_id": task_id,
        "attempt_number": attempt_number,
        "status": status,
        **fields,
    }


def execution_event(task_id: str, event_type: str, **fields: Any) -> dict[str, Any]:
    return {
        **_base("ExecutionEvent", fields.pop("id", None), fields.pop("created_at", None)),
        "task_id": task_id,
        "event_type": event_type,
        **fields,
    }


def artifact(uri: str, **fields: Any) -> dict[str, Any]:
    return {**_base("Artifact", fields.pop("id", None), fields.pop("created_at", None)), "uri": uri, **fields}


def result(task_id: str, status: str, **fields: Any) -> dict[str, Any]:
    return {
        **_base("Result", fields.pop("id", None), fields.pop("created_at", None)),
        "task_id": task_id,
        "status": status,
        **fields,
    }


def verification(subject_type: str, subject_id: str, verifier_id: str, method: str, outcome: str, **fields: Any) -> dict[str, Any]:
    return {
        **_base("Verification", fields.pop("id", None), fields.pop("created_at", None)),
        "subject": {"type": subject_type, "id": subject_id},
        "verifier_id": verifier_id,
        "method": method,
        "outcome": outcome,
        **fields,
    }


def work_receipt(task_id: str, result_id: str, status: str, **fields: Any) -> dict[str, Any]:
    return {
        **_base("WorkReceipt", fields.pop("id", None), fields.pop("created_at", None)),
        "task_id": task_id,
        "result_id": result_id,
        "status": status,
        **fields,
    }

File: sdk/typescript/src/v02.ts
export const PROTOCOL_VERSION = "0.2" as const;

export type Envelope = Record<string, unknown> & {
  protocol_version: "0.2";
  id: string;
  type: string;
  created_at: string;
};

function now(): string {
  return new Date().toISOString();
}

function id(): string {
  return globalThis.crypto?.randomUUID?.() ?? `dovewai-${Date.now()}-${Math.random().toString(16).slice(2)}`;
}

function base(type: string, fields: Record<string, unknown>): Envelope {
  const objectId = typeof fields.id === "string" ? fields.id : id();
  const createdAt = typeof fields.created_at === "string" ? fields.created_at : now();
  const rest = { ...fields };
  delete rest.id;
  delete rest.created_at;
  return { protocol_version: PROTOCOL_VERSION, id: objectId, type, created_at: createdAt, ...rest };
}

export function task(intent: string, fields: Record<string, unknown> = {}): Envelope {
  return base("Task", { ...fields, intent });
}

export function capability(name: string, version: string, fields: Record<string, unknown> = {}): Envelope {
  return base("Capability", { ...fields, name, version });
}

export function claim(task_id: string, holder_id: string, mode: "read" | "write", lease_expires_at: string, fields: Record<string, unknown> = {}): Envelope {
  return base("Claim", { ...fields, task_id, holder_id, mode, lease_expires_at });
}

export function attempt(task_id: string, attempt_number: number, status: "pending" | "running" | "succeeded" | "failed" | "cancelled", fields: Record<string, unknown> = {}): Envelope {
  return base("Attempt", { ...fields, task_id, attempt_number, status });
}

export function executionEvent(task_id: string, event_type: string, fields: Record<string, unknown> = {}): Envelope {
  return base("ExecutionEvent", { ...fields, task_id, event_type });
}

export function artifact(uri: string, fields: Record<string, unknown> = {}): Envelope {
  return base("Artifact", { ...fields, uri });
}

export function result(task_id: string, status: "succeeded" | "failed" | "cancelled", fields: Record<string, unknown> = {}): Envelope {
  return base("Result", { ...fields, task_id, status });
}

export function verification(subjectType: string, subjectId: string, verifier_id: string, method: string, outcome: "passed" | "failed" | "indeterminate" | "skipped", fields: Record<string, unknown> = {}): Envelope {
  return base("Verification", { ...fields, subject: { type: subjectType, id: subjectId }, verifier_id, method, outcome });
}

export function workReceipt(task_id: string, result_id: string, status: "recorded" | "verified" | "verification_failed" | "indeterminate", fields: Record<string, unknown> = {}): Envelope {
  return base("WorkReceipt", { ...fields, task_id, result_id, status });
}
